### DISCUSSION TITLE
xUnit theory method not covered

### DISCUSSION THREAD

**Discussion by kfrajtak** [SteveGilham/altcover#121] [Q&A]
I have noticed that coverage of some of the methods that I know are executed during the test is missing. These methods are called from a `[Theory]` test method:
```
public class ES008DALTests
{
	[Theory]
	[CsvDataSource(@"ES008DAL.pict.csv")]
	public void ES008DALTest(ES008Parameters parameters)
	{
	    var dal = new ES008DAL();
	    dal.GetDataSet(parameters);
	}
}
``` 
I can see "everything" being visited/covered/recorded in the file with the collected coverage - the `ES008DALTest` constructor, the custom `CsvDataSource` to read the data from a CSV file, the `ES008Parameters` constructor. But not the `ES008DALTest` method.

**Comment by SteveGilham** [comment]
The presence or not of a `[Theory]` attribute would only be significant if there was an explicit attribute filter.  The presence of some code generation to wire up the data source may have an effect on the actual IL generated, and what comes out the end of that process is what a coverage tool has to deal with.

There are a couple of ways of tying the coverage report with what is actually being analysed -- for example looking at what the IL code looks like when decompiled with a tool like e.g. [ILSpy](https://github.com/icsharpcode/ILSpy), and finding where the [Theory] decoration has ended up; or by reconciling the coverage data with the actual code by line and column e.g. using the [visualizer tool](https://www.nuget.org/packages/altcover.visualizer/) to paint the source with colours representing whether it's covered (blue), not covered (red) or not even considered for coverage (grey).

**Comment by kfrajtak** [comment]
* I can see the tests being executed in output window while running the `vstest.console.exe`. 
* I can see the AltCover instrumentation of said class and its methods in ILSpy.
* I can see the information about lines not being covered (rendered in red) and branches not being visited in the visualizer tool.

**Comment by SteveGilham** [comment]
That, to coin a phrase, is blinking well baffling.  If the above is a minimal repro, then, as this seems to involve a payware component, I'd be interested in examining the just instrumented test assembly and the associated XML, if at all possible.

**Comment by kfrajtak** [comment]
I will try to extract the MVP and share it with you.

**Reply by kfrajtak** [reply]
I tried to reproduce it outside the main testing process and it worked.
Which files would you like me to share with you?

**Reply by SteveGilham** [reply]
Ideally, the test assembly as source, as built, and as instrumented; the command line used to instrument the code, and the XML report.  That should be enough to do the necessary forensics on.

**Reply by kfrajtak** [reply]
![image](https://user-images.githubusercontent.com/1510452/106108065-0e46e980-6148-11eb-915c-6237b2679550.png)
The XML report file is 121 MB after instrumentation and another 128 MB after collection. 
I'll try to find out if I can share the assembly ...

**Reply by SteveGilham** [reply]
The XML compresses very nicely.  You'll need to throw everything into a zip archive to post here, anyway.

**Reply by kfrajtak** [reply]
This is the result of running
```
"C:\Utils\AltCover\Coverage\net472\AltCover.exe" --single -i=Tests.xUnit\bin\Debug -s=FluentAssertions -s=xunit -s=System -s=Tests -s=Microsoft -s=Ninject -s=Infragistics -s=Grape -s=ClosedXML -s=Connector -s=Humanizer -s=NCalc -s=OpenTracing -s=WCFExtras -s=GUI
```
[Uploading cov_output.zip…]()

**Reply by SteveGilham** [reply]
You can, of course, remove `<Module>` tags referring to actual business logic assemblies from the report first.

**Reply by kfrajtak** [reply]
[cov_output_xunit.zip](https://github.com/SteveGilham/altcover/files/5886705/cov_output_xunit.zip)

**Reply by SteveGilham** [reply]
For the test assembly, it's enough for diagnostic purposes to have a test assembly built with only the method that's not reporting and the minimum of other code around (which may mean you can cut the test assembly down to just the class posted at the top).

**Reply by SteveGilham** [reply]
Looking at the XML, is this the method you were looking for?  Starting at line 1061479 --
```
            <Method visited="false" cyclomaticComplexity="1" nPathComplexity="0" sequenceCoverage="0" branchCoverage="0" isConstructor="false" isStatic="false" isGetter="false" isSetter="false" crapScore="0">
              <Summary numSequencePoints="4" visitedSequencePoints="0" numBranchPoints="1" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="1" minCyclomaticComplexity="1" visitedClasses="0" numClasses="0" visitedMethods="0" numMethods="1" minCrapScore="0" maxCrapScore="0" />
              <MetadataToken>100664083</MetadataToken>
              <Name>System.Void Lekis.Lekis2005.LpWTests.xUnit.BLL.EconomicStatistics.ES008DALTests::ES008DALTest(Lekis.Lekis2005.BLL.EconomicStatistics.Parameters.ES008Parameters)</Name>
              <FileRef uid="3066" />
              <SequencePoints>
                <SequencePoint vc="0" uspid="8649" ordinal="0" offset="0" sl="15" sc="9" el="15" ec="10" bec="0" bev="0" fileid="3066" />
                <SequencePoint vc="0" uspid="8648" ordinal="1" offset="1" sl="16" sc="13" el="16" ec="38" bec="0" bev="0" fileid="3066" />
                <SequencePoint vc="0" uspid="8647" ordinal="2" offset="7" sl="17" sc="13" el="17" ec="40" bec="0" bev="0" fileid="3066" />
                <SequencePoint vc="0" uspid="8646" ordinal="3" offset="15" sl="18" sc="9" el="18" ec="10" bec="0" bev="0" fileid="3066" />
              </SequencePoints>
              <BranchPoints />
              <MethodPoint />
            </Method>
```
